File: church/report/archdeaconry_dashboard/archdeaconry_dashboard.js
frappe.query_reports["Archdeaconry Dashboard"] = {
    "filters": [
        {
            "fieldname": "diocese",
            "label": __("Diocese"),
            "fieldtype": "Link",
            "options": "Diocese",
            "get_query": function() {
                return {
                    filters: {
                        "is_group": 1
                    }
                };
            }
        },
        {
            "fieldname": "archdeaconry",
            "label": __("Archdeaconry"),
            "fieldtype": "Link",
            "options": "Archdeaconry",
            "get_query": function() {
                var diocese = frappe.query_report.get_filter_value('diocese');
                return {
                    filters: {
                        "diocese": diocese
                    }
                };
            },
            "depends_on": "diocese"
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        }
    ],
    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);
        if (column.fieldname == "total_contributions" && data && data.total_contributions) {
            value = "<span style='color:green'>" + value + "</span>";
        }
        if (column.fieldname == "avg_contribution_per_member" && data && data.avg_contribution_per_member) {
            value = "<span style='color:blue'>" + value + "</span>";
        }
        return value;
    },
    "tree": true,
    "name_field": "name",
    "parent_field": "parent",
    "initial_depth": 1,
    "onload": function(report) {
        report.page.add_inner_button(__("View on Map"), function() {
            frappe.set_route("gps-dashboard", {"entity_type": "Archdeaconry"});
        });
    },
    "get_datatable_options": function(options) {
        return Object.assign(options, {
            treeView: true,
            cellHeight: 45,
            columns: [
                { name: "name", width: 300 },
                { name: "type", width: 100 },
                { name: "total_members", width: 120 },
                { name: "total_contributions", width: 150 },
                { name: "avg_contribution_per_member", width: 180 }
            ]
        });
    },
    "get_chart_data": function(columns, result) {
        return {
            data: {
                labels: result.map(d => d.name),
                datasets: [
                    {
                        name: "Total Members",
                        values: result.map(d => d.total_members)
                    },
                    {
                        name: "Total Contributions",
                        values: result.map(d => d.total_contributions)
                    }
                ]
            },
            type: 'bar',
            colors: ['#7cd6fd', '#743ee2']
        }
    },
    "after_datatable_render": function(datatable_obj) {
        $(datatable_obj.wrapper).find(".dt-row-0").find('input[type=checkbox]').click();
    },
    "get_chart_options": function(columns, result) {
        return {
            title: {
                text: 'Archdeaconry Overview'
            },
            subtitle: {
                text: 'Members and Contributions'
            },
            xAxis: {
                title: {
                    text: 'Archdeaconry'
                }
            },
            yAxis: [
                {
                    title: {
                        text: 'Total Members'
                    }
                },
                {
                    title: {
                        text: 'Total Contributions'
                    },
                    opposite: true
                }
            ]
        }
    }
};

File: church/report/archdeaconry_dashboard/archdeaconry_dashboard.py
import frappe
from frappe import _

def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data)
    summary = get_summary(data)
    return columns, data, None, chart, summary

def get_columns():
    return [
        {
            "fieldname": "name",
            "label": _("Name"),
            "fieldtype": "Data",
            "width": 200
        },
        {
            "fieldname": "type",
            "label": _("Type"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "total_members",
            "label": _("Total Members"),
            "fieldtype": "Int",
            "width": 120
        },
        {
            "fieldname": "total_contributions",
            "label": _("Total Contributions"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "avg_contribution_per_member",
            "label": _("Avg Contribution/Member"),
            "fieldtype": "Currency",
            "width": 180
        }
    ]

def get_data(filters):
    data = []
    archdeaconries = get_archdeaconries(filters)
    
    for archdeaconry in archdeaconries:
        archdeaconry_data = get_entity_data("Archdeaconry", archdeaconry.name, filters)
        archdeaconry_data["parent"] = None
        data.append(archdeaconry_data)
        
        parishes = get_parishes(archdeaconry.name)
        for parish in parishes:
            parish_data = get_entity_data("Parish", parish.name, filters)
            parish_data["parent"] = archdeaconry.name
            data.append(parish_data)

    return data

def get_entity_data(doctype, name, filters):
    total_members = get_total_members(doctype, name)
    total_contributions = get_total_contributions(doctype, name, filters)
    avg_contribution = total_contributions / total_members if total_members else 0
    
    return {
        "name": name,
        "type": doctype,
        "total_members": total_members,
        "total_contributions": total_contributions,
        "avg_contribution_per_member": avg_contribution
    }

def get_archdeaconries(filters):
    archdeaconry_filters = {}
    if filters.get("diocese"):
        archdeaconry_filters["diocese"] = filters.get("diocese")
    if filters.get("archdeaconry"):
        archdeaconry_filters["name"] = filters.get("archdeaconry")
    return frappe.get_all("Archdeaconry", filters=archdeaconry_filters, fields=["name"])

def get_parishes(archdeaconry):
    return frappe.get_all("Parish", filters={"archdeaconry": archdeaconry}, fields=["name"])

def get_total_members(doctype, name):
    if doctype == "Archdeaconry":
        return frappe.db.count("Church Member", filters={"archdeaconry": name})
    elif doctype == "Parish":
        return frappe.db.count("Church Member", filters={"parish": name})
    return 0

def get_total_contributions(doctype, name, filters):
    date_filter = []
    if filters.get("from_date"):
        date_filter.append(["payment_date", ">=", filters.get("from_date")])
    if filters.get("to_date"):
        date_filter.append(["payment_date", "<=", filters.get("to_date")])

    if doctype == "Archdeaconry":
        members = frappe.get_all("Church Member", filters={"archdeaconry": name}, pluck="name")
    elif doctype == "Parish":
        members = frappe.get_all("Church Member", filters={"parish": name}, pluck="name")
    else:
        return 0

    if not members:
        return 0

    total = frappe.db.sql("""
        SELECT SUM(amount) 
        FROM `tabChurch Payment`
        WHERE member IN (%s) AND docstatus = 1
        %s
    """ % (', '.join(['%s'] * len(members)), 
           "AND " + " AND ".join([f"{f[0]} {f[1]} %s" for f in date_filter]) if date_filter else ""),
    tuple(members + [f[2] for f in date_filter]))

    return total[0][0] if total and total[0][0] else 0

def get_chart_data(data):
    labels = [d['name'] for d in data if d['type'] == 'Archdeaconry']
    contributions = [d['total_contributions'] for d in data if d['type'] == 'Archdeaconry']
    members = [d['total_members'] for d in data if d['type'] == 'Archdeaconry']

    return {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": "Total Contributions",
                    "values": contributions
                },
                {
                    "name": "Total Members",
                    "values": members
                }
            ]
        },
        "type": "bar",
        "colors": ["#4CAF50", "#2196F3"]
    }

def get_summary(data):
    archdeaconry_data = [d for d in data if d['type'] == 'Archdeaconry']
    total_members = sum(d['total_members'] for d in archdeaconry_data)
    total_contributions = sum(d['total_contributions'] for d in archdeaconry_data)
    avg_contribution = total_contributions / total_members if total_members else 0

    return [
        {
            "value": total_members,
            "label": _("Total Members"),
            "datatype": "Int",
            "currency": None
        },
        {
            "value": total_contributions,
            "label": _("Total Contributions"),
            "datatype": "Currency",
            "currency": frappe.defaults.get_global_default("currency")
        },
        {
            "value": avg_contribution,
            "label": _("Average Contribution per Member"),
            "datatype": "Currency",
            "currency": frappe.defaults.get_global_default("currency")
        }
    ]